### ISSUE TITLE
Rule for missing PEP 698 `@override` decorators?

### ISSUE THREAD

**Issue by guillp** [astral-sh/ruff#6844]
An idea for a new rule: checking that methods in subclasses that override methods from the parent are decorated with `@override`.
This decorator is defined in [PEP 698](https://peps.python.org/pep-0698/) and is meant to help both linters and developpers to recognize methods that are overriding methods from a parent class.

Example:

```python

class A:
    def foo(self) -> None:
        print("foo from class A!")

class B(A):
    def foo(self) -> None:
        print("foo from class B!")
```

A rule would detect that the `@override` decorator is missing on `B.foo()` since it overrides `A.foo()`.

Note that until python 3.12, `override` is only available from `typing-extensions`.

**Comment by zanieb** [comment]
TIL!

This sounds good to me as a great Python 3.12 milestone although I do not think it should be enabled by default.

**Comment by charliermarsh** [comment]
The challenge here is that we won’t be able to enforce this except for cases in which the parent class and subclass are defined in the same file (until we have multi-file analysis).

**Comment by zanieb** [comment]
Great point, it may not be worth prioritizing until then since there will be many false negatives

**Comment by JonathanPlasse** [comment]
There is also the inverse problem of having `@override` on a class without bases, or `@overload` on a function without the function without `@overload`.

**Comment by aaronsteers** [comment]
I found this thread while looking for the same. Are there any other linters that currently support this? 

Asked also here:

- https://github.com/mkorpela/overrides/issues/120

**Comment by illusional** [comment]
Is this something that might be more appropriate for red-knot?

**Comment by folkvir** [comment]
up! 🫡  Is there is any work on this since then? We're also debating on how to properly define subclassed methods and having a rule about enforcing the usage of `@override` would be very appreciated!

**Comment by nickbroon** [comment]
FWIW, MyPy has an [`--enable-error-code explicit-override` option][1] that checks this. 

[1]: https://mypy.readthedocs.io/en/stable/error_code_list2.html#check-that-override-is-used-when-overriding-a-base-class-method-explicit-override

**Comment by MichaReiser** [comment]
I consider it very likely that we'll add a rule like this to ty in the near future.

**Comment by AlexWaygood** [comment]
> I consider it very likely that we'll add a rule like this to ty in the near future.

It's the second bullet point (the only not-yet-done thing) in https://github.com/astral-sh/ty/issues/155

**Comment by Wesztman** [comment]
Stumbled upon missing this rule the other day, looking forward to getting it merged into ty 😁🎉
